common/lib/host_list_provider/monitoring/cluster_topology_monitor.ts

```diff
@@ -0,0 +1,482 @@
+  private untrackedPromises: Promise<void>[] = [];
+
+  // Signals to other methods that asynchronous operations have completed/should be completed.
+  private requestToUpdateTopology: boolean = false;
+  private releaseTopologyUpdate: () => void;
+  private topologyUpdated = new Promise<void>((done) => {
+    this.releaseTopologyUpdate = () => {
+      done();
+    };
+  });
+
+  constructor(
+    clusterId: string,
+    topologyMap: CacheMap<string, HostInfo[]>,
+    initialHostSpec: HostInfo,
+    props,
+    pluginService: PluginService,
+    hostListProvider: MonitoringRdsHostListProvider,
+    refreshRateMs: number,
+    highRefreshRateMs: number
+  ) {
+    this.clusterId = clusterId;
+    this.topologyMap = topologyMap;
+    this.initialHostInfo = initialHostSpec;
+    this._pluginService = pluginService;
+    this._hostListProvider = hostListProvider;
+    this._properties = props;
+    this.refreshRateMs = refreshRateMs;
+    this.highRefreshRateMs = highRefreshRateMs;
+    this.runPromise = this.run();
+  }
+
+  get hostListProvider(): MonitoringRdsHostListProvider {
+    return this._hostListProvider;
+  }
+  get pluginService(): PluginService {
+    return this._pluginService;
+  }
+  get properties(): Map<string, any> {
+    return this._properties;
+  }
+
+  async close(): Promise<void> {
+    this.stopMonitoring = true;
+    this.hostMonitorsStop = true;
+    await Promise.all(this.runPromise ? this.untrackedPromises.concat(this.runPromise) : this.untrackedPromises);
+    this.untrackedPromises = [];
+    this.hostMonitors.clear();
+  }
+
+  async forceMonitoringRefresh(shouldVerifyWriter: boolean, timeoutMs: number): Promise<HostInfo[]> {
+    const currentHosts = this.topologyMap.get(this.clusterId);
+    if (currentHosts) {
+      return currentHosts;
+    }
+    if (shouldVerifyWriter) {
+      const client = this.monitoringClient;
+      this.monitoringClient = null;
+      this.isVerifiedWriterConnection = false;
+      await this.closeConnection(client);
+    }
+
+    return await this.waitTillTopologyGetsUpdated(timeoutMs);
+  }
+
+  async forceRefresh(client: any, timeoutMs: number): Promise<HostInfo[]> {
+    if (this.isVerifiedWriterConnection) {
+      return await this.waitTillTopologyGetsUpdated(timeoutMs);
+    }
+
+    // Otherwise use provided unverified connection to update topology
+    return await this.fetchTopologyAndUpdateCache(client);
+  }
+
+  async waitTillTopologyGetsUpdated(timeoutMs: number): Promise<HostInfo[]> {
+    const currentHosts: HostInfo[] = this.topologyMap.get(this.clusterId);
+    let latestHosts: HostInfo[];
+
+    // Notify monitor (that might be in delay) that topology should be updated.
+    this.requestToUpdateTopology = true;
+
+    if (timeoutMs === 0) {
+      return currentHosts;
+    }
+
+    let timeoutId: any;
+    const timeoutTask: Promise<void> = new Promise((resolve, reject) => {
+      timeoutId = setTimeout(() => {
+        reject("Topology update timed out.");
```

Comment:
Instead of reject lets throw a custom error here similar to failover
```suggestion
        reject(new TopologyUpdateTimeoutError());
```

Reply:
Actually, on second thought, instead of writing a new timeoutTask, can we just reuse the `getTimeoutTask()` in utils.ts